implementations/rust/ockam/ockam_core/src/access_control/cache.rs

```diff
@@ -0,0 +1,351 @@
+use crate::compat::sync::Mutex;
+use crate::{
+    Address, IncomingAccessControl, LocalInfo, OutgoingAccessControl, RelayMessage, Route,
+};
+use alloc::vec::Vec;
+use async_trait::async_trait;
+use core::fmt::Debug;
+use std::time::Instant;
+
+const CACHE_MAX_SIZE: usize = 10;
+const CACHE_DURATION_SECS: u64 = 1;
+
+#[derive(Debug)]
+struct CacheEntry {
+    source: Address,
+    destination: Address,
+    onward_route: Route,
+    return_route: Route,
+    local_info: Vec<LocalInfo>,
+    timestamp: Instant,
+}
+
+impl CacheEntry {
+    fn from(relay_message: &RelayMessage) -> Self {
+        Self {
+            source: relay_message.source().clone(),
+            destination: relay_message.destination().clone(),
+            onward_route: relay_message.onward_route().clone(),
+            return_route: relay_message.return_route().clone(),
+            local_info: relay_message.local_message().local_info(),
+            timestamp: Instant::now(),
+        }
+    }
+
+    /// Returns true if the cache entry is expired.
+    fn is_expired(&self) -> bool {
+        self.timestamp.elapsed().as_secs() >= CACHE_DURATION_SECS
+    }
+
+    /// Returns true if the relay message matches the cache entry.
+    /// Everything except the payload is compared.
+    fn matches(&self, relay_message: &RelayMessage) -> bool {
+        self.source == *relay_message.source()
+            && self.destination == *relay_message.destination()
+            && self.onward_route == *relay_message.onward_route()
+            && self.return_route == *relay_message.return_route()
+            && self.local_info == relay_message.local_message().local_info_ref()
+    }
+}
+
+#[derive(Debug)]
+struct Cache {
+    cache: Mutex<Vec<CacheEntry>>,
+}
+
+impl Cache {
+    pub fn new() -> Self {
+        Self {
+            cache: Mutex::new(Vec::new()),
+        }
+    }
+
+    /// Returns true if the relay message is in the cache and not expired.
+    pub fn exist_in_cache(&self, relay_message: &RelayMessage) -> bool {
+        let mut cache_guard = self.cache.lock().unwrap();
+        let position = cache_guard
+            .iter()
+            .position(|entry| entry.matches(relay_message));
+        if let Some(position) = position {
+            let entry = cache_guard.get_mut(position).unwrap();
+            if entry.is_expired() {
+                cache_guard.remove(position);
+                false
+            } else {
+                true
+            }
+        } else {
+            false
+        }
+    }
+
+    /// Adds the relay message to the cache.
+    pub fn add_authorized(&self, relay_message: &RelayMessage) {
+        let mut cache_guard = self.cache.lock().unwrap();
+        let position = cache_guard
+            .iter()
+            .position(|entry| entry.matches(relay_message));
+        if let Some(position) = position {
+            cache_guard.remove(position);
+        }
```

Comment:
Just for readability, I'd write:

```rust
cache_guard
    .iter()
    .position(|entry| entry.matches(relay_message))
    .map(|position| cache_guard.remove(position));
```

Reply:
Having side effects on `map()` without reading the output is dangerous, since it could be lazy and never be evaluated (and it's already lazy on iterators). Ideally, `inspect()` would be a better alternative, but it's [not stable yet ](https://github.com/rust-lang/rust/issues/91345).

Reply:
Oh right! I didn't try to compile my proposal, but makes sense that is not a good idea because we are iterating and removing from the same source right? I don't think it even compiles. Good to know about that `inspect` function, never seen it before.